src/library/basesqltablemodel.cpp

```diff
@@ -334,16 +334,16 @@ void BaseSqlTableModel::select() {
              << m_rowInfo.size();
 }
 
-void BaseSqlTableModel::setTable(const QString& tableName,
-        const QString& idColumn,
-        const QStringList& tableColumns,
+void BaseSqlTableModel::setTable(QString tableName,
+        QString idColumn,
+        QStringList tableColumns,
         QSharedPointer<BaseTrackCache> trackSource) {
     if (sDebug) {
         qDebug() << this << "setTable" << tableName << tableColumns << idColumn;
     }
-    m_tableName = tableName;
-    m_idColumn = idColumn;
-    m_tableColumns = tableColumns;
+    m_tableName = std::move(tableName);
+    m_idColumn = std::move(idColumn);
+    m_tableColumns = std::move(tableColumns);
 
```

Comment:
Your https://godbolt.org/z/TEvY5e7ff does not give the correct impression, because the structure is empty and the compile cannot optimize the functions away because of the message output and the disabled optimization.
I have played with a more realistic https://godbolt.org/z/sdGz5nYbW
Here is turns out that the copy and rvalue function have the same assembler code, while the original const ref function creates a deep copy. With the copy version, the caller is free to copy or move the value in. That's why I prefer the copy version.  

Finally it depends also weather the object is cheap to move or not. In case of QT object with the Pimpl pattern moving is just passing a pointer. So even if there is an extra move for some reasons, it does not matter but keeps the interface simple.

Reply:
I did a second look why the value and the lref version are equal. It is bacuse copying an lref is also copying a pointer. 
So if moving is also only a matter of a pointer copy from an inlined move constructor, it is the same. 

In this case here the lref version requires a copy for some library l features in the calling code. To avoid that, we would need a const ref overload as well. That's why the copy version is more universal with the cost of an extra move in cases where it cannot inlined.

Reply:
> Your https://godbolt.org/z/TEvY5e7ff does not give the correct impression, because the structure is empty and the compile cannot optimize the functions away because of the message output and the disabled optimization.

Not really, the fact that there are user defined copy-/move-constructors at all inhibited optimization (since the type is no longer trivially move-/copyable). Looking at the assembly all of the calls still get inlined. Regardless, that was not the point I was trying to make. I was rather talking about the semantics present in C++ which get enforced by the language. 

>  I have played with a more realistic https://godbolt.org/z/sdGz5nYbW

In the case of Qt-classes this example is even more unrealistic IMO because it relies on the optimizer. For the optimizer to do its job it relies on being able to see the source, which is only the case for move assignment and the move constructor for many Qt classes (see the [move assign and move ctor in Qstring header](https://codebrowser.dev/qt5/qtbase/src/corelib/text/qstring.h.html#_ZN7QStringC1EOS_). 

Anyways, I still don't understand why one should not take by `&&` when there is an unconditional move in the body. If the caller really doesn't want the value to be moved out of, they can explicitly copy it into a temporary later. That way you only get a single move by default and if you really need the copy, the caller can create it explicitly beforehand.

Reply:
> I was rather talking about the semantics present in C++ which get enforced by the language.

From that point of view your conclusion is correct.

> Anyway, I still don't understand why one should not take by && when there is an unconditional move in the body.

For me it is a good rule of thumb to use copy by value in case the function keeps the value and use const ref otherwise. This allows all things like: 
```
A a;
function(a)
function(move(a))
function(otherFunctionReturingA())
```
Which results in a bare minimum of deep copies and works with one function signature. 
This works of cause only optimal if `A::A(A&& a)` is equal cheap compared to copy an address.     
Otherwise you need to take care and implement the `const ref` version and the `non const lref` overload like you have proposed as universal solution.

Reply:
I disagree with your conclusion. The proposed solution of taking `A&&` does not need any overloads, uses the same amount of copies and even fewer moves. The only difference is that the caller then explicitly needs to specify whether to copy or to move (which I actually prefer over the default value semantics). This demonstrates the phenomenon:
https://godbolt.org/z/YEexffYbn

Reply:
In details for the const ref. If you want to call a rvalue-only function with a const ref, you need to copy explicitly (good imo) while if you do the same for a function taking by value, the copy is implicit. If you do the same for a non-const lref, the compiler still complains, forcing you to either make an explicit copy or explicit move. required implicit moves are still performed while implicit copies are eliminated.

Reply:
True, however that would be an unusual style.

Reply:
Fair. Though I'd be very keen to know why since I don't see any downsights to this style.

Reply:
This is just because of legacy.

Reply:
Can you elaborate? What's there to say against using the rvalue-ref-only solution?

Reply:
rvalue references is a "new" c++11 feature, and not a first class citizen like in Rust. When using them from a lvalue, `std::move(lvalue)` is mandatory. Requiring a mandatory `T(lvalue)` for a function would be a new concept, exposing an unnecessarily and implementation detail. Normally you see a `function(A a)` with a `function(const A& a)` to hide that from the calling code. Just providing `function(A&& a)` is breaking the normal `function(a)` call. The user is forced to consider if he wants to keep that value or throw it away. This decision should be optional only for those who know what they are doing.

We had a similar discussion some time ago that leads to this rule: 
https://github.com/mixxxdj/mixxx/wiki/Coding-Guidelines#r-value-references--move-constructors--move-assignment-operators

Some details and reasoning can be found here:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#fcall-parameter-passing

Reply:
Mhmm, thank you for these resources. I agree that having to specify whether to move or copy is a breaking change and also not super easy to understand with the usual obtuse error messages generated by c++ compilers. That said, for this case, even the core guidelines recommend taking by it by `&&` universally for this particular case where we just unconditionally move. You can see that in the "Advanced Parameter passing" figure in the "in + move" case. This is explained in more details in [F.18  For “will-move-from” parameters, pass by X&& and std::move the parameter](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f18-for-will-move-from-parameters-pass-by-x-and-stdmove-the-parameter). There is an exception for "Unique owner types that are move-only and cheap-to-move, such as unique_ptr" but I don't think Qt's implicitly-shared classes really qualify for that since they're not types that are only concerned with ownership.

Reply:
Taking a && is an advanced technik where the guide is clear about: 

> Use the advanced techniques only after demonstrating need, and document that need in a comment.

We require the same in the Mixxxx code guiline.

Reply:
I acknowledge that. I'm fine with that as long as the Type in question is cheap to move (which includes the implicitly-shared Qt containers). Is that ok with you?

Reply:
Yes. Maybe after that long discussion we need to update the single sentences in the Mixxx guideline.